Book & User test

joredudiaz opened the pull request:
# Summary
- book and user model with test
# Trello card
https://trello.com/c/TGTFgN8w

joredudiaz committed:
UPDATE Gemfile/Gemfile.lock gems for test (database_cleaner, factory_bot and faker)

joredudiaz committed:
UPDATE
Gemfile/Gemfile.lock
shoulda, shoulda-matchers, factory_bot_rails,  rspec-rails

spec/factories/*.rb
factories for test

spec/models/*_spec.rb
test for models

joredudiaz committed:
UPDATE
rubocop fix

holywyvern commented on spec/models/book_spec.rb:
```diff
@@ -0,0 +1,45 @@
+  it { is_expected.to validate_presence_of(:title) }
+  it { is_expected.to validate_presence_of(:editor) }
+  it { is_expected.to validate_presence_of(:year) }
+  it {
```
Multiline blocks should always be created using `do...end` blocks.

holywyvern commented on spec/models/book_spec.rb:
```diff
@@ -0,0 +1,45 @@
+      false
+    end
+  }
+  it {
```
Same as last block. One liners are ok to be in between curly brackets

holywyvern commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,18 @@
+  it { is_expected.to validate_presence_of(:first_name) }
+  it { is_expected.to validate_presence_of(:last_name) }
+
+  #   describe '#generate_verification_code' do
```
Remove commented code.

joredudiaz committed:
UPDATE
spec/models/book_spec.rb
change block to do..end instruction

spec/models/user_spec.rb
remove unnecessary comments

leoncalermo commented on spec/models/book_spec.rb:
```diff
@@ -0,0 +1,45 @@
+  it { is_expected.to validate_presence_of(:editor) }
+  it { is_expected.to validate_presence_of(:year) }
+  it do
+    expect(book.year).to satisfy do |v|
```
why you didn't use the the numericality validate method ? 🤔  
something like this `validate_numericality_of(:year).only_integer`

joredudiaz committed:
change validations of length and check if it's only numbers

leoncalermo commented on spec/factories/user_factory.rb:
```diff
@@ -0,0 +1,13 @@
+    first_name { Faker::Name.first_name }
+    last_name { Faker::Name.last_name }
+    email { Faker::Internet.email }
+    password { password }
```
instead of having the variable password you can do like this
```
password { Faker::Internet.password(8) }
password_confirmation { password }
```

leoncalermo commented on spec/factories/user_factory.rb:
```diff
@@ -0,0 +1,13 @@
+FactoryBot.define do
+  factory :user, class: User do
+    password = Faker::Internet.password(8)
+
```
and delete this blank line 👀

leoncalermo commented on spec/factories/book_factory.rb:
```diff
@@ -0,0 +1,12 @@
+    image { Faker::Simpsons.quote }
+    title { Faker::Book.title }
+    editor { Faker::Book.publisher }
+    year { Faker::Date.between(2.days.ago, Time.zone.today).to_s[0, 4] }
```
I find something weird about this,  because, in the validation you said that should validate that is an integer value, but here you are setting the year as a string

leoncalermo commented on spec/models/book_spec.rb:
```diff
@@ -0,0 +1,35 @@
+  it { is_expected.to validate_length_of(:year).is_equal_to(4) }
+end
+
+describe Book do
```
I think that this whole describe is doing the same validation that in the 6th line, because you are validating that should be valid in the 6th line, the part of checking that a new record was added  after creation, should be in the controller, that it's not in the scope for this PR,

leoncalermo commented on Gemfile:
```diff
@@ -63,6 +63,21 @@ group :test do
+  gem 'database_cleaner'
+
+  # Fake data to test
+  gem 'faker'
```
you maaaay want to change the faker and the factory_bots_rails to the `group :development, :test`, believe me, in development you will be able to create records  in the Rails Console that could be very helpful to tests thing on your own

joredudiaz committed:
change faker and factory_bot_rails gems into development and test environment, remove variable password and put data field password into password_confirmation, remove unnecesary test

joredudiaz committed:
Merge branch 'book' into book_test
